### ISSUE TITLE
Is it possible to increase the order of the biquad filter?

### ISSUE THREAD

**Issue by Hozeeben** [aubio/aubio#269]
Like the title mention I wan't to increase the order of the biquad filter so the slope of the cutoff frequency gets steeper.
I looked through the files but could'nt find something.
Is it possible that i overlooked something or is there some workaround for this?

**Comment by piem** [comment]
hello @Hozeeben

i think you are looking for `filter`, implemented in `src/temporal/filter.h`, and which is named `digital_filter` in the python module to avoid namespace collision with the python keyword.

have a look at `python/demos/demo_filter.py` for an example in python, and `tests/src/temporal/test-filter.c` for one in C.

I'll let you close that issue if that answers your question.

best, piem

**Comment by Hozeeben** [comment]
Do you mean this command from `python/demos/demo_filter.py` to change the order: 
`f = aubio.digital_filter(7)`
If yes i get following error if i change the value
```
f.set_biquad(1.31235898*10**-4, 2.62471796*10**-4, 1.31235898*10**-4, -1.98642633, 0.98695127) 
ValueError: error when setting filter with biquad coefficients
```
That is the code related to it:
   ```
 f = aubio.digital_filter(5)
    # f.set_a_weighting(samplerate)
    f.set_biquad(1.31235898*10**-4, 2.62471796*10**-4, 1.31235898*10**-4, -1.98642633, 0.98695127)
```
The base value in `aubio.digital_filter` is `3`

**Comment by piem** [comment]
oh i see, a function is missing to set arbitrary coefficients.

I'm preparing a branch adding a `digital_filter.set_coeffs(b, a)` function, will upload in a bit.

**Comment by piem** [comment]
@Hozeeben could you check the branch `feature/filter_set_coeffs` and see if that works for you?

**Comment by Hozeeben** [comment]
Can you give me a little code excample for that because i don't really understand the parameters you give into that funktion.
I've tried it with some values but the spectrogram of the two outputfiles is'nt different.

Right now I use the Code from `aubio/python/demos/demo_filter.py` but instead of an A-weighted filter i use a biquad filter.
The biquad values are `set_biquad(1.00187537, -1.98008047, 0.97848147, -1.98012083, 0.98031648)` and the Song i use is KIRA New world from [NoCopyrightSounds](http://ncs.io/NewWorld)
Just in case you wanna try it yourself (:

**Comment by piem** [comment]
with `order = 3, f = aubio.digital_filter(order)` and the new `set_coeffs`:

    f.set_biquad(1.00187537, -1.98008047, 0.97848147, -1.98012083, 0.98031648)

is equivalent to:

    b = np.zeros(order, dtype="float64")
    a = np.zeros(order, dtype="float64")
    b[0] =  1.00187537
    b[1] = -1.98008047
    b[2] =  0.97848147
    a[0] =  1.0
    a[1] = -1.98012083
    a[2] =  0.98031648
    f.set_coeffs(b, a)

a few things are a bit confusing here:
- `order` is actually `N+1` in the [classic notation](https://en.wikipedia.org/wiki/Digital_filter#Characterization),
- `a[0]` is always 1,
- the type of `a` and `b` should be `float64` (this is tricky since

**Comment by Hozeeben** [comment]
It works perfectly for me (:
Thank you for the fast response (:

**Comment by theronic** [comment]
@Hozeeben the order of a filter refers to the number of poles in the filter. By definition, a biquad filter is a [second order filter](https://en.wikipedia.org/wiki/Digital_biquad_filter): two poles, two zeros. The only way to get a higher order filter out of a biquad filter is to serially cascade multiple biquad sections, potentially with a single-order stage. So those coefficients will be for chained filters, but I'm not sure that's what's happening here. Those don't look like enough coefficients to me, unless they are multiple filters with duplicated coefficients.

**Comment by Hozeeben** [comment]
@theronic Yeah i searched for a way to do it after i relised that i dont know how exactly i have to calculate the coefficients.
I tested your answer and it worked just fine so i have a sharp slope now (:

Never the less the function that got implemented should be interesting for iir filter of an higher order that could be relevant for others (:

**Comment by piem** [comment]
hi @theronic, and thanks for your additional input.

you will have noticed that the example i gave was precisely how to use `set_coeffs` on a biquad filter: in aubio `order = 3` is equivalent to a 2nd order filter. The name of this parameter is misleading, but it matches the length of `b` vector.

As for higher order filters, there are examples in aubio with the A-weighting and C-weighting filters.

@Hozeeben i'm re-opening this issue since i have a few more commits to push to that branch. in particular, I would like to add tests the float vs. double in python. thanks for testing it out.

best, piem
